llvm/lib/CodeGen/SpillPlacement.h

```diff
@@ -38,13 +39,21 @@ class BitVector;
 class EdgeBundles;
 class MachineBlockFrequencyInfo;
 class MachineFunction;
+class SpillPlacementWrapperLegacy;
+class SpillPlacementAnalysis;
+
+class SpillPlacement {
+  friend class SpillPlacementWrapperLegacy;
+  friend class SpillPlacementAnalysis;
 
-class SpillPlacement : public MachineFunctionPass {
   struct Node;
+
   const MachineFunction *MF = nullptr;
   const EdgeBundles *bundles = nullptr;
   const MachineBlockFrequencyInfo *MBFI = nullptr;
-  Node *nodes = nullptr;
+
+  static void arrayDeleter(Node *N);
+  std::unique_ptr<Node, decltype(&arrayDeleter)> nodes;
```

Comment:
Array form would be better:
```suggestion
  std::unique_ptr<Node[]> nodes;
```
https://en.cppreference.com/w/cpp/memory/unique_ptr

Reply:
The definition of `Node` is not available here, so the default deleter fails to compile sizeof(Node) for this incomplete type. To hack around it I put the definition of `arrayDeleter` in the implementation where struct Node is defined.

But changing to `unique_ptr<Node[], theDeleter>` facilitates removal of `.get()` calls

Reply:
An outlined default destructor would work. 🤔

Reply:
That works